README.md
# SCS 智能车仿真平台

搬运自[SourceForge项目](https://sourceforge.net/p/smartcarsim/)，已获得作者[yukunlinykl]授权

以全国大学生智能车竞赛为背景，
依照智能车竞赛规则建立仿真模型，旨在为广大参与智能车
竞赛的参赛选手提供一个仿真智能车机械，验证软件控制算
法的平台。

该平台以刚体模型库 OpenDE 和图形库 OpenGL 为基础，
描述并绘制智能车以及赛道的机械模型，仿真智能车行进的
物理过程。


## 版本信息

本发布包为 SCS 的源代码发布包，版本号为 1.0.1 。

本发布包包含SCS-1.0.1的源文件、头文件，ode、freeglut
的库文件、头文件，以及 Linux、VS2008、VS2010三个环境
下的样例工程。样例工程中包含三个组别的样例 AI 决策函
数，以及样例绘制函数。


## 目录结构及说明
```
SCS-1.0.1	根目录
├── scs			SCS 目录
│   ├── include			SCS 头文件目录
│   └── src			SCS 源文件目录
├── ai			样例 AI 目录
│   ├── include			样例 AI 头文目录
│   ├── src			样例 AI 源文目录
│   └── track			赛道文件目录
├── build		编译目录
│   ├── include			编译所需头文件目录
│   ├── linux			Linux 环境编译目录
│   ├── VS2008			VS2008 环境编译目录
│   └── VS2010			VS2010 环境编译目录
├── doc			文档目录
│   ├── INSTALL.TXT		使用方法说明文件
│   ├── LICENSE.TXT		用户协议文件
│   ├── NEWS.TXT		开发新闻文件
│   ├── ChangeLog.TXT		修改日志文件
│   ├── SCS-1.0.1.pdf		使用手册及编程接口
│   └── src			文档源代码目录
└── README.md		本文件
```

### 完整的目录树
```
SCS-1.0.1
├── ai
│   ├── build
│   │   ├── ai
│   │   ├── api.o
│   │   ├── car.o
│   │   ├── draw.o
│   │   ├── libscs.a
│   │   ├── main.o
│   │   ├── Makefile
│   │   ├── simulation.o
│   │   ├── track.o
│   │   ├── ui.o
│   │   └── vector.o
│   ├── include
│   │   ├── scs.h
│   │   └── vector.h
│   ├── src
│   │   └── main.cpp
│   └── track
│       ├── final8.trk
│       └── province8.trk
├── build
│   ├── include
│   │   ├── GL
│   │   │   ├── freeglut_ext.h
│   │   │   ├── freeglut.h
│   │   │   ├── freeglut_std.h
│   │   │   └── glut.h
│   │   └── ode
│   │       ├── collision.h
│   │       ├── collision_space.h
│   │       ├── collision_trimesh.h
│   │       ├── common.h
│   │       ├── compatibility.h
│   │       ├── contact.h
│   │       ├── error.h
│   │       ├── export-dif.h
│   │       ├── mass.h
│   │       ├── matrix.h
│   │       ├── memory.h
│   │       ├── misc.h
│   │       ├── objects.h
│   │       ├── odeconfig.h
│   │       ├── odecpp_collision.h
│   │       ├── odecpp.h
│   │       ├── ode.h
│   │       ├── odeinit.h
│   │       ├── odemath.h
│   │       ├── odemath_legacy.h
│   │       ├── rotation.h
│   │       └── timer.h
│   ├── linux
│   │   ├── ai
│   │   ├── api.o
│   │   ├── car.o
│   │   ├── draw.o
│   │   ├── libscs.a
│   │   ├── main.o
│   │   ├── Makefile
│   │   ├── simulation.o
│   │   ├── track.o
│   │   ├── ui.o
│   │   └── vector.o
│   ├── VS2008
│   │   ├── ai
│   │   │   ├── ai.vcproj
│   │   │   └── lib
│   │   │       ├── freeglut_static.lib
│   │   │       └── ode_double.lib
│   │   ├── scs
│   │   │   └── scs.vcproj
│   │   ├── VS2008.sln
│   │   └── VS2008.suo
│   └── VS2010
│       ├── ai
│       │   ├── ai.vcxproj
│       │   ├── ai.vcxproj.filters
│       │   └── lib
│       │       ├── freeglut_static.lib
│       │       └── ode_double.lib
│       ├── scs
│       │   ├── scs.vcxproj
│       │   └── scs.vcxproj.filters
│       ├── VS2010.sln
│       └── VS2010.suo
├── doc
│   ├── ChangeLog.TXT
│   ├── INSTALL.TXT
│   ├── NEWS.TXT
│   ├── SCS-1.0.1.pdf
│   └── src
│       ├── avatar
│       │   ├── andy.png
│       │   ├── berniw.jpg
│       │   ├── fengzi.png
│       │   ├── free.jpg
│       │   ├── hanquan.jpg
│       │   ├── may.jpg
│       │   ├── Remi-Coulom.jpg
│       │   └── tju_speed.jpg
│       ├── figure
│       │   ├── coordinate_balance.png
│       │   ├── coordinate.png
│       │   ├── demo1.png
│       │   ├── demo2.png
│       │   ├── demo3.png
│       │   ├── logic.pdf
│       │   ├── time.pdf
│       │   └── window.png
│       ├── Makefile
│       └── SCS-1.0.1.tex
├── LICENSE
├── README.md
└── scs
    ├── include
    │   ├── api.h
    │   ├── car.h
    │   ├── common.h
    │   ├── draw.h
    │   ├── key.h
    │   ├── mytime.h
    │   ├── simulation.h
    │   ├── track.h
    │   ├── ui.h
    │   └── vector.h
    └── src
        ├── api.cpp
        ├── car.cpp
        ├── draw.cpp
        ├── simulation.cpp
        ├── track.cpp
        ├── ui.cpp
        └── vector.cpp

25 directories, 108 files
```

## 使用方法

请先阅读 doc/ 目录下的用户协议 LICENSE.TXT ，然后阅读
同目录下的 INSTALL.TXT 文件。


## 联系方式

作者邮箱	yukunlinykl@users.sf.net

中文主页	http://www.smartcarsim.com/

英文主页	http://sourceforge.net/p/smartcarsim/


doc/NEWS.TXT
2013年9月1日
在www.smartcarsim.com发布修订版 SCS-1.0.1

2013年8月31日
申请域名www.smartcarsim.com用于SCS的发布

2013年8月1日
在SourceForge发布正式版 SCS-1.0

2013年7月1日
源代码迁移至SourceForge

2013年5月1日
发布beta版公测，版本号为SCS-0.99

2013年2月1日
在智能车制作论坛发布alpha版公测，版本号为SCS-0.9

2012年11月1日
在智能车制作论坛发布pre-alpha版内测，版本号为SCS-0.2

2012年10月1日
在学校内部发布内测版，命名为SCS

2012年8月1日
开始着手开发SCS



scs/src/draw.cpp
/*
 *	VERSION:	SCS-1.0.1
 *	DATE:		2013.Sep.1
 *	AUTHOR:		Yu Kunlin
 */

#include <stdio.h>
#include <time.h>

#include <GL/glut.h>
#include <ode/ode.h>

#include "mytime.h"
#include "simulation.h"
#include "key.h"
#include "car.h"
#include "track.h"
#include "common.h"
#include "draw.h"
#include "api.h"
#include "ui.h"

	// window handle
	int WinCustom[CUSTOMWINDOWNUM],WinCar,WinGod;

	ViewType viewtype = bird;
	DriveMode drivemode = ai;
	CarType cartype = camera;
	int RouteFlag = 0;

	// these three var discribe the location of view camera
	double pitch = 45;
	double height = 2.0;
	double h = 0.0;
	static double x=0.0, y=0.0;
	double distance = 1.0;

	// simulate speed of virtual world
	double VirtualSpeed = 1.0;

	int CustomWindowFlag = 0;
	int CustomWindowNum = 0;
	int PathFlag = 0;

	double hpr[3] = {0,30,0};
	sVector ViewPoint(-0.5,0,0.5);

	sVector CameraPos(0.0,0.0,0.3);
	double DepressionAngle = 30;

static void DrawGround ()
{
	glLineWidth (1);
	glColor3d (0.2f,0.2f,0.2f);
	for (int i=-20; i<=20; i++) {	// vertical line
		glBegin (GL_LINES);
		glVertex3d (i,-20,0);
		glVertex3d (i, 20,0);
		glEnd ();
	}
	for (int i=-20; i<=20; i++) {	// horizental line
		glBegin (GL_LINES);
		glVertex3d (-20,i,0);
		glVertex3d ( 20,i,0);
		glEnd ();
	}
}

static void SetViewPoint (double hpr[3], sVector ViewPoint)
{
	glLoadIdentity ();
	// mult inverse matrix of camera place to current matrix
	glRotated (hpr[2],0,0,1);
	glRotated (hpr[1],1,0,0);
	glRotated (hpr[0],0,1,0);
	glRotated (-90,1,0,0);
	glRotated ( 90,0,0,1);
	glTranslated (-ViewPoint.GetX(),-ViewPoint.GetY(),-ViewPoint.GetZ());
}

void ClearViewVar ()
{
	pitch = 45;
	height = 2.0;

	h = 0.0;
	x=0.0;
	y=0.0;
	distance = 1.0;

	ViewPointDelta.set (0.0,0.0,0.0);
}

static void GodView ()	// calc camera place to view car
{
	sVector pos(dBodyGetPosition (Chassis->body));
	sVector lv(dBodyGetLinearVel (Chassis->body));
	sVector dir = CarY ();
	static double speed = 1.0;

	switch (viewtype) {
		case over:	// `~
			hpr[0] = TrackReverseFlag * 180.0;
			hpr[1] = 90.0;
			hpr[2] = 0.0;
			ViewPoint.SetX (ViewPoint.GetX()*5.0/10.0 + pos.GetX()*5.0/10.0);
			ViewPoint.SetY (ViewPoint.GetY()*7.0/10.0 + pos.GetY()*3.0/10.0);
			ViewPoint.SetZ (height * distance*2.0+pos.GetZ());
			break;
		case car:	// 1
			hpr[0] = -dir.DirDeg();
			hpr[1] = DepressionAngle - asin(dir.GetZ())*RAD_TO_DEG;
			hpr[2] = 0.0;
			ViewPoint = ToWorldCoo(CameraPos) + pos;
			break;
		case bird:	// 2
			hpr[0] = TrackReverseFlag * 180.0;
			hpr[1] = pitch;
			hpr[2] = 0.0;
			speed -= 0.5;
			speed = speed*29.0/30.0 + lv.GetLen()/30.0;
			speed += 0.5;
			ViewPoint.SetX (ViewPoint.GetX()*7.0/10.0 + pos.GetX()*3.0/10.0 - cos((hpr[0])*DEG_TO_RAD)*distance*speed/5.0);
			ViewPoint.SetY (ViewPoint.GetY()*7.0/10.0 + pos.GetY()*3.0/10.0 + sin((hpr[0])*DEG_TO_RAD)*distance*speed/5.0);
			ViewPoint.SetZ (height * distance*speed/3.0+pos.GetZ());
			break;
		case birdf:	// 3
			hpr[0] -= lv.GetX()*sin(hpr[0]*DEG_TO_RAD)*VirtualSpeed+lv.GetY()*cos(hpr[0]*DEG_TO_RAD)*VirtualSpeed;
			hpr[1] = pitch;
			hpr[2] = 0.0;
			speed -= 0.5;
			speed = speed*29.0/30.0 + lv.GetLen()/30.0;
			speed += 0.5;
			ViewPoint.SetX (ViewPoint.GetX()*7.0/10.0 + pos.GetX()*3.0/10.0 - cos((hpr[0])*DEG_TO_RAD)*distance*speed/5.0);
			ViewPoint.SetY (ViewPoint.GetY()*7.0/10.0 + pos.GetY()*3.0/10.0 + sin((hpr[0])*DEG_TO_RAD)*distance*speed/5.0);
			ViewPoint.SetZ (height * distance*speed/3.0+pos.GetZ());
			break;
		case hard:	// 4
			hpr[0] = -dir.DirDeg();
			hpr[1] = pitch;
			hpr[2] = 0.0;
			ViewPoint.SetX (pos.GetX() - cos((hpr[0])*DEG_TO_RAD)*distance);
			ViewPoint.SetY (pos.GetY() + sin((hpr[0])*DEG_TO_RAD)*distance);
			ViewPoint.SetZ (height * distance+pos.GetZ());
			break;
		case back:	// 5
			hpr[0] = -dir.DirDeg();
			hpr[1] = pitch/2.0;
			hpr[2] = 0.0;
			ViewPoint.SetX (pos.GetX() - cos((hpr[0])*DEG_TO_RAD)*distance/3.0);
			ViewPoint.SetY (pos.GetY() + sin((hpr[0])*DEG_TO_RAD)*distance/3.0);
			ViewPoint.SetZ (height * distance/8.0+pos.GetZ());
			break;
		case overf:	// 6
			hpr[0] = -dir.DirDeg();
			hpr[1] = 90.0;
			hpr[2] = 0.0;
			ViewPoint = pos;
			ViewPoint.SetZ (height * distance*2.0+pos.GetZ());
			break;
		case soft:	// 7
			hpr[0] -= lv.GetX()*sin(hpr[0]*DEG_TO_RAD)*6+lv.GetY()*cos(hpr[0]*DEG_TO_RAD)*6;
			hpr[1] = pitch;
			hpr[2] = 0.0;
			ViewPoint.SetX (pos.GetX() - cos((hpr[0])*DEG_TO_RAD)*distance);
			ViewPoint.SetY (pos.GetY() + sin((hpr[0])*DEG_TO_RAD)*distance);
			ViewPoint.SetZ (height * distance+pos.GetZ());
			break;

		case mouse:	// 0
			hpr[0] = 0;
			hpr[1] = pitch;
			ViewPoint.SetZ(height);
			break;
	}
	if (viewtype!=car) {
		hpr[0] += h;
		ViewPoint += ViewPointDelta/100.0;
	}
	SetViewPoint (hpr,ViewPoint);
	if (viewtype!=car) {
		hpr[0] -= h;
		ViewPoint -= ViewPointDelta/100.0;
	}
}

static void CarView ()	// set the place of the camera(the camera of the car, not godview camera)
{
	double hpr[3];  
	sVector pos (dBodyGetPosition (Chassis->body));
	sVector dir = CarY() ;
	sVector ViewPoint = ToWorldCoo(CameraPos) + pos;

	hpr[0] = -dir.DirDeg();
	hpr[1] = DepressionAngle - asin(dir.GetZ())*RAD_TO_DEG;
	hpr[2] = 0.0;

	SetViewPoint (hpr,ViewPoint);
}

static void DrawRoute ()
{
	int Length = 5000;
	sRouteID p,q;
	static sRoute car,car_old;
	sVector posl(dBodyGetPosition (Wheel_BL->body));
	sVector posr(dBodyGetPosition (Wheel_BR->body));
	if (VirtualSpeed>0.0) {
		p = head;
		head = (sRouteID)malloc(sizeof(sRoute));
		head->pos = (posl+posr)/2.0;
		head->next = p;
		if (p)
			head->speed = (head->pos - p->pos).GetLen()/VirtualSpeed;
	}
	p = head;
	int i=0;
	glLineWidth(1);
	glBegin (GL_LINE_STRIP);
	while (p->next) {
		q = p->next;
		glColor3d (1.0-i/1000.0,1.0-i/1000.0,1.0-i/1000.0);
		//double v = p->speed*20;
		//glColor3d (v,v,v);
		glVertex3d (p->pos.GetX(),p->pos.GetY(),p->pos.GetZ());
		p = q;
		if (i++>Length) {
			Free (p);
			break ;
		}
	}
	glEnd ();
}

static void DrawSpeed ()
{
	glLoadIdentity ();
	glTranslated (0,0,-5);
	glColor3d (0.9f,0.2f,0.2f);
	double width = 2.7;
	double height = -1.5;
	double RulerWidth = 0.4;

	// draw ruler
	for (int i=-2; i<8; i++) {
		if (i%5==0)	glLineWidth (3);
		else		glLineWidth (1);
		glBegin (GL_LINES);
		glVertex2d (i%5==0?width-RulerWidth-0.1:width-RulerWidth,i/2.0+height);
		glVertex2d (width,i/2.0+height);
		glEnd ();
	}
	glBegin (GL_LINES);
	glVertex2d (width,height-2);
	glVertex2d (width,height+4);
	glEnd ();

	// draw speed
	sVector lv(dBodyGetLinearVel (Chassis->body));

	int temp[2];				// max line widths are different between OSes
	glGetIntegerv(GL_LINE_WIDTH_RANGE,temp);
	if (temp[1]>10) temp[1] = 10;
	glLineWidth ((GLfloat)temp[1]);

	glBegin (GL_LINES);
	glVertex2d (width,height);
	glVertex2d (width,lv.GetLen()/2+height);
	glEnd ();

	static double duty = 0.0;
	duty = duty*0.9 + (MotorDutyL + MotorDutyR)/40.0*0.1;
	glColor3d (0.2f,0.2f,0.9f);
	glBegin (GL_LINES);
	glVertex2d (width-0.2,height);
	glVertex2d (width-0.2,duty+height);
	glEnd ();

	static double speed = 0.0;
	speed = speed*0.9 + sGetSpeed()*0.1;
	glColor3d (0.2f,0.9f,0.2f);
	glBegin (GL_LINES);
	glVertex2d (width-0.1,height);
	glVertex2d (width-0.1,speed+height);
	glEnd ();
}

static void DrawDirection ()
{
	static double dir = ServoDir/2.0;
	glLoadIdentity ();
	glTranslated (0,-1.5,-5);
	
	dir = dir*8.0/10.0+ServoDir/10.0;
	glColor3d (0.2f,0.2f,0.9f);
	glLineWidth (1);
	glBegin (GL_LINES);
	glVertex2d (0,-0.05);
	glVertex2d (0, 0.02);
	glEnd ();

	glLineWidth (10);
	glBegin (GL_LINES);
	glVertex2d (0,0);
	glVertex2d (dir*4.0/100.0*(CarDirection?-1:1),0);
	glEnd ();
}

static void DrawCCDLine ()
{
	glLoadIdentity ();
	glTranslated (0.0,0.0,-1.0);
	glBegin (GL_LINES);
		glVertex2d (10.0,0.0);
		glVertex2d (-10.0,0.0);
	glEnd ();
}

static void DrawGodWindow ()		// main window, you can use every viewtype
{
	glutSetWindow (WinGod);
	if (viewtype==car)
		glClearColor (0.66f,0.66f,0.66f,1.0f);
	else	glClearColor (0.3f,0.3f,0.3f,1.0f);
	glClear (GL_COLOR_BUFFER_BIT);

	GodView ();
	DrawGround ();

	glEnable (GL_DEPTH_TEST);
	glClear (GL_DEPTH_BUFFER_BIT);
	glDepthFunc (GL_LEQUAL);

	DrawTrack ();
	if (viewtype == car) goto finish;

	if (RouteFlag)
		DrawRoute ();
	Free (NULL);

	switch (cartype) {
		case camera:		DrawCar();		break;
		case electromagnetic:	DrawCar();		break;
		case balance:		DrawBalanceCar ();	break;
	}
finish:
	glDisable (GL_DEPTH_TEST);
	if (cartype==balance && viewtype==car)
		DrawCCDLine ();
	DrawSpeed ();
	if (cartype!=balance)
		DrawDirection ();

	glFlush ();
	glutSwapBuffers ();
}
static void DrawCarWindow ()		// mini window, only can use car viewtype
{
	glutSetWindow (WinCar);

	glClearColor (0.66f,0.66f,0.66f,1.0f);
	glClear (GL_COLOR_BUFFER_BIT);

	CarView ();
	DrawGround ();
	
	glEnable (GL_DEPTH_TEST);
	glClear (GL_DEPTH_BUFFER_BIT);
	glDepthFunc (GL_LEQUAL);

	DrawTrack ();
	glDisable (GL_DEPTH_TEST);
	if (cartype==balance)
		DrawCCDLine ();


	glFlush ();
	glutSwapBuffers ();
}


static void Ai ()	// call AI call back function
{
	static int first = 1;
	Ip = 0;
	if (AiFunc)	AiFunc ();
	else if (first) {
		printf ("AI: There is no AI function.\n");
		drivemode = play;
		first = 0;
	}
}

static void DrawCustomWindow ()
{
	for (int i=0; i<CustomWindowNum; i++) {
		glutSetWindow (WinCustom[i]);
		glClear (GL_COLOR_BUFFER_BIT);
		glLoadIdentity (); 
		glColor3d (1.0,1.0,1.0);

		if (DisplayFunc[i])
			DisplayFunc[i] ();

		glFlush ();
		glutSwapBuffers ();
	}
}

static void callback(int value)		// MAIN FRAME OF TOTAL PROGRAM!!!!!
{
	glutTimerFunc(FIELDTIME_M,&callback,0);

	static double ControlTime = 0.0;
	double fTimes = FIELDTIME/STEPTIME;
	int iTimes = (int)fTimes;
	double tail = fTimes - iTimes;

	DrawGodWindow ();

	if (drivemode==play || drivemode==debug) Play ();

	for (int i=0; i<iTimes*VirtualSpeed; i++) {
		step (STEPTIME);			// simulate
		ControlTime += STEPTIME;
		if (ControlTime>=CONTROLTIME) {
			ControlTime -= CONTROLTIME;
			DrawCarWindow ();
			if (drivemode!=play)	Ai ();	// auto dive (according to CarWindow)
			if (CustomWindowFlag)
						DrawCustomWindow ();
		}
	}
	if (tail>0.00001) {
		step (tail*VirtualSpeed);
		ControlTime += tail;
		if (ControlTime>=CONTROLTIME) {
			ControlTime -= CONTROLTIME;
		 	DrawCarWindow ();
			if (drivemode!=play)	Ai ();
			if (CustomWindowFlag)
						DrawCustomWindow ();
		}
	}
}

void DrawInit()
{
	glutInitDisplayMode (GLUT_RGB | GLUT_DOUBLE | GLUT_DEPTH);

	// God Window
	int WindowWidthGod = WINDOW_WIDTH_GOD_0;
	int WindowHeightGod = WINDOW_HEIGHT_GOD_0;
	if (CustomWindowNum>2) {
		WindowWidthGod = WINDOW_WIDTH_GOD_1;
		WindowHeightGod = WINDOW_HEIGHT_GOD_1;
	}
	if (CustomWindowNum>4) {
		WindowWidthGod = WINDOW_WIDTH_GOD_2;
		WindowHeightGod = WINDOW_HEIGHT_GOD_2;
	}
	glutInitWindowSize (WindowWidthGod,WindowHeightGod);
	glutInitWindowPosition (WINDOW_POS_X,WINDOW_POS_Y);
	WinGod = glutCreateWindow (VERSION);
	
	glutKeyboardFunc (&Key);
	glutSpecialFunc (&SpecialKeyPress);
	glutSpecialUpFunc (&SpecialKeyUp);
	glutMouseFunc (&Mouse);
	glutMotionFunc (&motion);
	glutReshapeFunc (&Reshape);

	glMatrixMode (GL_PROJECTION);
	gluPerspective (45.0f,(GLdouble)WindowWidthGod/(GLdouble)WindowHeightGod,0.01f,100.0f);
	glMatrixMode (GL_MODELVIEW);


	// Car Window
	glutInitWindowSize (WINDOW_WIDTH_CAR,WINDOW_HEIGHT_CAR);
	glutInitWindowPosition (WINDOW_POS_X+WindowWidthGod+20,WINDOW_POS_Y);
	WinCar = glutCreateWindow ("Car View");

	glutKeyboardFunc (&Key);
	glutSpecialFunc (&SpecialKeyPress);
	glutSpecialUpFunc (&SpecialKeyUp);
	glutReshapeFunc (&Reshape);

	glMatrixMode (GL_PROJECTION);
	gluPerspective (45.0f,4.0/3.0,0.01f,100.0f);
	glMatrixMode (GL_MODELVIEW);

	// Custom Window
	if (CustomWindowFlag)
	if (CustomWindowNum>4) for (int i=0; i<CustomWindowNum; i++) {
		int x = (i+2)%3;
		int y = (i+2)/3;
		char name[] = "Custom Window 00";
		glutInitWindowSize (WINDOW_WIDTH_CUSTOM,WINDOW_HEIGHT_CUSTOM);
		glutInitWindowPosition (WINDOW_POS_X + x*(WINDOW_WIDTH_CUSTOM+20),WINDOW_POS_Y + (y)*WINDOW_HEIGHT_CUSTOM);
		sprintf(name,"Custom Window %d",i);
		WinCustom[i] = glutCreateWindow (name);	

		glutKeyboardFunc (&Key);
		glutSpecialFunc (&SpecialKeyPress);
		glutSpecialUpFunc (&SpecialKeyUp);
	}
	else if (CustomWindowNum>2) for (int i=0; i<CustomWindowNum; i++) {
		int x = (i+5)%3;
		int y = (i+5)/3;
		char name[] = "Custom Window 00";
		glutInitWindowSize (WINDOW_WIDTH_CUSTOM,WINDOW_HEIGHT_CUSTOM);
		glutInitWindowPosition (WINDOW_POS_X + x*(WINDOW_WIDTH_CUSTOM+10),WINDOW_POS_Y + (y)*WINDOW_HEIGHT_CUSTOM);
		sprintf(name,"Custom Window %d",i);
		WinCustom[i] = glutCreateWindow (name);

		glutKeyboardFunc (&Key);
		glutSpecialFunc (&SpecialKeyPress);
		glutSpecialUpFunc (&SpecialKeyUp);
	}
	else for (int i=0; i<CustomWindowNum; i++) {
		char name[] = "Custom Window 00";
		glutInitWindowSize (WINDOW_WIDTH_CUSTOM,WINDOW_HEIGHT_CUSTOM);
		glutInitWindowPosition (WINDOW_POS_X+WindowWidthGod+20,WINDOW_POS_Y + (i+1)*WINDOW_HEIGHT_CUSTOM);
		sprintf(name,"Custom Window %d",i);
		WinCustom[i] = glutCreateWindow (name);

		glutKeyboardFunc (&Key);
		glutSpecialFunc (&SpecialKeyPress);
		glutSpecialUpFunc (&SpecialKeyUp);
	}
	printf("GUI: Glut initialized!\n");

	// All Window have one same timer callback function
	// And it's the main frame of whole program
	glutTimerFunc(FIELDTIME_M,&callback,0);
}
